server/engine/src/main/java/org/eclipse/lsp/cobol/service/DocumentModelService.java

```diff
@@ -78,87 +80,91 @@ public void processAnalysisResult(String uri, AnalysisResult analysisResult) {
+                    d -> {
+                      d.setOpened(false);
+                      removeAllRelatedDiagnostics(d);
+                    });
   }
 
   private void removeAllRelatedDiagnostics(CobolDocumentModel documentModel) {
     Optional.ofNullable(documentModel.getAnalysisResult())
-        .map(AnalysisResult::getDiagnostics)
-        .ifPresent(
-            d ->
-                d.forEach(
-                    (uri, closedFileRelatedDiagnostics) -> {
-                      List<Diagnostic> oldDiagnostics = diagnosticRepo.get(uri);
-                        List<Diagnostic> newDiagnostics = Optional.ofNullable(oldDiagnostics)
-                                .orElse(Collections.emptyList())
-                                .parallelStream()
-                                .filter(diagnosticsNotContains(closedFileRelatedDiagnostics))
-                                .collect(Collectors.toList());
-                        diagnosticRepo.put(ImmutableMap.of(uri, newDiagnostics));
-                    }));
+            .map(AnalysisResult::getDiagnostics)
+            .ifPresent(
+                    d ->
+                            d.forEach(
+                                    (uri, closedFileRelatedDiagnostics) -> {
+                                      List<Diagnostic> oldDiagnostics = diagnosticRepo.get(uri);
+                                      List<Diagnostic> newDiagnostics = Optional.ofNullable(oldDiagnostics)
+                                              .orElse(Collections.emptyList())
+                                              .parallelStream()
+                                              .filter(diagnosticsNotContains(closedFileRelatedDiagnostics))
+                                              .collect(Collectors.toList());
+                                      diagnosticRepo.put(uri, newDiagnostics);
+                                    }));
   }
 
-    private Predicate<Diagnostic> diagnosticsNotContains(List<Diagnostic> closedFileRelatedDiagnostics) {
-      // diagnostics address location is matched and not the value as the same diagnostics could be generated by diff files.
-        return dia -> closedFileRelatedDiagnostics.stream().noneMatch(d1 -> d1 == dia);
-    }
+  private Predicate<Diagnostic> diagnosticsNotContains(List<Diagnostic> closedFileRelatedDiagnostics) {
+    // diagnostics address location is matched and not the value as the same diagnostics could be generated by diff files.
+    return dia -> closedFileRelatedDiagnostics.stream().noneMatch(d1 -> d1 == dia);
+  }
 
   /**
    * Removes document from registry
+   *
    * @param uri - document uri
    */
   @Synchronized
   public void removeDocument(String uri) {
     Optional.ofNullable(docs.get(uri))
-        .ifPresent(
-            d -> {
-              diagnosticRepo.delete(uri);
-              docs.remove(uri);
-            });
+            .ifPresent(
+                    d -> {
+                      diagnosticRepo.remove(uri);
```

Comment:
Should we update the repo with empty diagnostics instead of removing? diagnostics for the opened program seems to remain, but all related diagnostics disappear.

Reply:
The idea here is that an empty list of diagnostics means no errors in the file, but null (missing key) means the absent of information about the URI. At least I think it does :)
Do you have a specific case where there is an issue with diagnostic handling?

Reply:
I tried below scenario for diagnostics

1. Open a Cobol program with an error (or introduce an error).
2. Close the  program.
3. Expect all diagnostics are removed. I think in current scenario, diagnostics remains even if we close the program.
.

Reply:
I think I've fixed it, could you try again, please?

Reply:
works !